src/Pate/Verification.hs

```diff
@@ -1196,49 +1157,112 @@ validRegister ::
-        MT.llvmPtrEq sym ptrO (PSR.macawRegValue entry)
-      Nothing -> return $ W4.truePred sym
+      Just blockStart -> do
+        ptrO <- PD.concreteToLLVM blockStart
+        return $ macawRegBinding entry (PSR.ptrToEntry ptrO)
+      Nothing -> return $ mempty
     RegSP -> do
       stackRegion <- CMR.asks envStackRegion
       let
         CLM.LLVMPointer region _ = PSR.macawRegValue entry
-      liftIO $ W4.isEq sym region stackRegion
+      return $ exprBinding region stackRegion
     RegBV -> liftIO $ do
       let
         CLM.LLVMPointer region _ = PSR.macawRegValue entry
       zero <- W4.natLit sym 0
-      W4.isEq sym region zero
+      return $ exprBinding region zero
     RegTOC -> do
-      let
-        CLM.LLVMPointer region val = PSR.macawRegValue entry
       globalRegion <- CMR.asks envGlobalRegion
-      validRegion <- liftIO $ W4.isEq sym region globalRegion
       (tocO, tocP) <- getCurrentTOCs
       tocBV <- case W4.knownRepr :: WhichBinaryRepr bin of
         OriginalRepr -> wLit tocO
         PatchedRepr -> wLit tocP
-      tocLit <- liftIO $ W4.isEq sym tocBV val
-      liftIO $ W4.andPred sym validRegion tocLit
-    _ -> return $ W4.truePred sym
+      let targetToc = CLM.LLVMPointer globalRegion tocBV
+      return $ macawRegBinding entry (PSR.ptrToEntry targetToc)
+    _ -> return $ mempty
 
 
 validInitState ::
   Maybe (PatchPair arch) ->
   SimState sym arch Original ->
   SimState sym arch Patched ->
-  EquivM sym arch (W4.Pred sym)
-validInitState mpPair stO stP = withSym $ \sym -> do
-  preds <- zipRegStates (simRegs stO) (simRegs stP) $ \r vO vP -> do
+  EquivM sym arch (AssumptionFrame sym)
+validInitState mpPair stO stP = do
+  fmap mconcat $ zipRegStates (simRegs stO) (simRegs stP) $ \r vO vP -> do
     validO <- validRegister (fmap pOrig mpPair) vO r
     validP <- validRegister (fmap pPatched mpPair) vP r
-    liftIO $ W4.andPred sym validO validP
-  liftIO $ allPreds sym preds
+    return $ validO <> validP
+
+-- | Reads from immutable data have known results.
+-- We collect all the reads that occurred during the trace, and then
+-- assert that those values are necessarily equivalent to the concrete
+-- value from the binary
+validConcreteReads ::
+  forall bin sym arch.
+  KnownBinary bin =>
+  SimOutput sym arch bin ->
+  EquivM sym arch (AssumptionFrame sym)
+validConcreteReads stOut = withSym $ \sym -> do
+  binCtx <- getBinCtx @bin
+  let
+    binmem = MBL.memoryImage $ binary binCtx
+
+    go :: Seq.Seq (MT.MemOp sym (MM.ArchAddrWidth arch)) -> EquivM sym arch (AssumptionFrame sym)
+    go mops = do
+      flatops <- fmap F.toList $ liftIO $ MT.flatMemOps sym mops
+      fmap mconcat $ mapM readConcrete flatops
+
+    readConcrete ::
+      MT.MemOp sym (MM.ArchAddrWidth arch) ->
+      EquivM sym arch (AssumptionFrame sym)
+    readConcrete (MT.MemOp (CLM.LLVMPointer reg off) dir _ sz val end) = do
+      case (W4.asNat reg, W4.asBV off, dir) of
+        (Just 0, Just off', MT.Read) -> do
```

Comment:
Does this mean that we only pull concrete values from region 0?

Reply:
Currently, yes. All addresses from the binary are flattened into absolute addresses at region 0, and so this is the only place we would expect to find concrete values.